### PR TITLE
SplinesLinearSolver new API

### LINKED ISSUE #443: Refactor existing `Matrix` and `Matrix_Sparse`
Following discussion #425

### PR DESCRIPTION
**Pull request by blegouix**
Follows #425
Closes #443

### CODE REVIEW

File: include/ddc/kernels/splines/splines_linear_solver_sparse.hpp [hunk 1]
```diff
@@ -172,30 +178,30 @@ class Matrix_Sparse : public Matrix
...
 
-    virtual double get_element([[maybe_unused]] int i, [[maybe_unused]] int j) const override
+    virtual double get_element([[maybe_unused]] std::size_t i, [[maybe_unused]] std::size_t j)
+            const override
     {
-        throw std::runtime_error("MatrixSparse::get_element() is not implemented because no API is "
-                                 "provided by Ginkgo");
+        throw std::runtime_error(
+                "SplinesLinearProblemSparse::get_element() is not implemented because no API is "
+                "provided by Ginkgo");
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_solver_sparse.hpp:186]:
```suggestion
        return m_matrix_dense->at(i, j);
```

Would this work ?

File: include/ddc/kernels/splines/splines_linear_solver_sparse.hpp [hunk 2]
```diff
@@ -114,16 +114,22 @@ unsigned int default_preconditionner_max_block_size() noexcept
...
  */
 template <class ExecSpace>
-class Matrix_Sparse : public Matrix
+class SplinesLinearProblemSparse : public SplinesLinearProblem<ExecSpace>
 {
-    using matrix_sparse_type = gko::matrix::Csr<double, int>;
+public:
+    using typename SplinesLinearProblem<ExecSpace>::MultiRHS;
+    using SplinesLinearProblem<ExecSpace>::size;
+    using SplinesLinearProblem<ExecSpace>::operator<<;
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_solver_sparse.hpp:129]:
```suggestion
```

Are these two aliases required to make it work ?

File: tests/splines/matrix.cpp [hunk 3]
```diff
@@ -52,14 +55,16 @@ void check_inverse(ddc::DSpan2D matrix, ddc::DSpan2D inv)
         for (std::size_t j(0); j < N; ++j) {
             double id_val = 0.0;
             for (std::size_t k(0); k < N; ++k) {
-                id_val += matrix(i, k) * inv(j, k);
+                id_val += matrix(i, k) * inv(k, j);
```
Comment (tpadioleau) [inline tests/splines/matrix.cpp:58]: Was it working because the matrix is symmetric ?

Comment (blegouix) [reply tests/splines/matrix.cpp:58]: The matrix being symmetric makes it work regardless of the choice of convention indeed. However the change here is aligned to the change of convention brought with the current MR: the elements of a matrix `A_ij` are identified with indices `(i, j)` (it was previously `(j,i)`)

File: include/ddc/kernels/splines/splines_linear_solver_sparse.hpp (hunk 2 above)
Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_solver_sparse.hpp:129]: `ddc/include/ddc/kernels/splines/splines_linear_solver_sparse.hpp(202): error: identifier "size" is undefined`

File: include/ddc/kernels/splines/splines_linear_solver_sparse.hpp (hunk 1 above)
Comment (blegouix) [reply include/ddc/kernels/splines/splines_linear_solver_sparse.hpp]: Yes, now that lifecycle is clearly defined it makes sense. I propose commit 6033f08

File: tests/splines/matrix.cpp (hunk 3 above)
Comment (tpadioleau) [reply tests/splines/matrix.cpp:58]: Do you mean `A_ij` was identified by `get/set_element(j, i, a_ij)` ? I am not sure about that

File: include/ddc/kernels/splines/splines_linear_solver_sparse.hpp (hunk 1 above)
Comment (tpadioleau) [reply include/ddc/kernels/splines/splines_linear_solver_sparse.hpp]: Yes good to me

File: include/ddc/kernels/splines/splines_linear_solver_sparse.hpp (hunk 2 above)
Comment (tpadioleau) [reply include/ddc/kernels/splines/splines_linear_solver_sparse.hpp:129]: I see, without this alias we would need to call `this->size()`. Let's keep the alias.

File: tests/splines/matrix.cpp (hunk 3 above)
Comment (blegouix) [reply tests/splines/matrix.cpp:58]: No sorry, not the matrix A, the multi-rhs b.

ie. here https://github.com/CExA-project/ddc/blob/main/include/ddc/kernels/splines/matrix.hpp#L122

Comment (blegouix) [reply tests/splines/matrix.cpp:58]: Call here: https://github.com/CExA-project/ddc/pull/444/files#diff-deb6768d88f9dcadb445a3de0cfff4717aef343c81c34c4f3bff6ae4e3653117L115

Comment (tpadioleau) [reply tests/splines/matrix.cpp:58]: I agree

File: include/ddc/kernels/splines/splines_linear_solver_sparse.hpp [hunk 4]
```diff
@@ -172,30 +178,28 @@ class Matrix_Sparse : public Matrix
         m_matrix_sparse = matrix_sparse_type::create(gko_exec, gko::dim<2>(mat_size, mat_size));
     }
 
-    virtual double get_element([[maybe_unused]] int i, [[maybe_unused]] int j) const override
+    virtual double get_element([[maybe_unused]] std::size_t i, [[maybe_unused]] std::size_t j)
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_solver_sparse.hpp:181]:
```suggestion
    virtual double get_element(std::size_t i, std::size_t j)
```

File: include/ddc/kernels/splines/splines_linear_solver_sparse.hpp [hunk 5]
```diff
@@ -172,30 +178,28 @@ class Matrix_Sparse : public Matrix
         m_matrix_sparse = matrix_sparse_type::create(gko_exec, gko::dim<2>(mat_size, mat_size));
     }
 
-    virtual double get_element([[maybe_unused]] int i, [[maybe_unused]] int j) const override
+    virtual double get_element(std::size_t i, std::size_t j)
+            const override
```
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_solver_sparse.hpp:182]:
```suggestion
    virtual double get_element(std::size_t i, std::size_t j) const override
```

**Pull request merged**